### PR TITLE
1228 multiroot interpreter ui changes

### PR DESCRIPTION
**Pull request by DonJayamanne**
Changes to workflow of configuring an interpreter in multiroot scenarios.
Includes global retry for tests (3 times)
(some refactoring and moving around of files and code cleanup)

### CODE REVIEW

File: src/client/common/contextKey.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    private lastValue: boolean;
+
+    constructor(private name: string) { }
+
+    public async set(value: boolean): Promise<void> {
+        if (this.lastValue === value) {
+            return;
+        }
+        this.lastValue = value;
+        await commands.executeCommand('setContext', this.name, this.lastValue);
```
Comment (brettcannon) [inline src/client/common/contextKey.ts]: Is `setContext` an Electron/Chrome thing?

File: src/client/interpreter/configuration/pythonPathUpdaterService.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import { WorkspacePythonPath } from '../contracts';
+import { IPythonPathUpdaterService, IPythonPathUpdaterServiceFactory } from './types';
+
+export class PythonPathUpdaterService {
+    constructor(private pythonPathSettingsUpdaterFactory: IPythonPathUpdaterServiceFactory) { }
+    public async updatePythonPath(pythonPath: string, configTarget: ConfigurationTarget, wkspace?: Uri): Promise<void> {
+        const pythonPathUpdater = this.getPythonUpdaterService(configTarget, wkspace);
+
+        try {
+
```
Comment (brettcannon) [inline src/client/interpreter/configuration/pythonPathUpdaterService.ts]: Extraneous newline.

File: src/client/interpreter/configuration/setInterpreterProvider.ts [hunk 3]
```diff
@@ -54,63 +51,64 @@ export class SetInterpreterProvider implements Disposable {
...
     }
-    private async presentQuickPick() {
-        const targetConfig = await this.getWorkspacePythonPath();
-        const resourceUri = targetConfig ? targetConfig.folderUri : undefined;
-        const suggestions = await this.getSuggestions(resourceUri);
+
+    private async getSuggestions(resourceUri?: Uri) {
+        const interpreters = await this.interpreterManager.getInterpreters(resourceUri);
+        // tslint:disable-next-line:no-non-null-assertion
+        interpreters.sort((a, b) => a.displayName! > b.displayName! ? 1 : -1);
```
Comment (brettcannon) [inline src/client/interpreter/configuration/setInterpreterProvider.ts]: Makes me miss the decorate-sort-undecorate pattern from Python.

File: src/client/common/contextKey.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/client/common/contextKey.ts]: vscode thingy

**Pull request merged**